### DISCUSSION TITLE
No garbage collection between parametrized test instances.

### DISCUSSION THREAD

**Discussion by Vert53** [pytest-dev/pytest#8153] [General] [answered]
Using Pytest version: 3.8.0 on ubuntu 18 and python3.6

I encountered an issue where a parametrized test uses objects which define a destructor (__del__) that is needed to be called between test instances of the same test function. I noticed that the garbage collector doesn't collect these objects unless I explicitly del them. I assume this might be because of some inner behavior of pytest, but this is highly misleading and has potential for problems.

```python
class Resource:
    def __init__(self):
        acquire_resource()
    def __del__(self):
        release_resource()


@pytest.mark.parametrize('param', (1,2,3))
def test_resource:
    res = Resource()
```

I except res to be collected when it is out of the scope of the function but it is not unless we call:
del res

Please let me know if this behavior is expected for some reason I missed.
If it is not perhaps we should fix it in the future

**Comment by RonnyPfannschmidt** [comment]
It's expected, use context managers

**Comment by asottile** [comment, marked as answer]
if you're using `__del__` for resource freeing you're depending on borderline-undefined-cpython-implementation-details.  as @RonnyPfannschmidt indicates you should adjust your api to use a context manager

**Comment by nicoddemus** [comment]
Just to complement the answers: this is not pytest specific, but how Python behaves; you will get the same behavior if you call `test_resource()` manually without using pytest at all.

**Comment by Vert53** [comment]
Thanks for the answers. Due to this being caused by garbage collector working in the background I found a possible workaround that might work:
Using the python gc module,[ gc.collect(](https://docs.python.org/3/library/gc.html#gc.collect)) can be run in the beginning of the test to block until the gc collects all objects from previous runs.
Do you think this is a good idea? If it is, perhaps some kind of "collect" flag could be added to ensure this happens. 
Just a note, the reason in this case resource clean up is critical is due to working with a C++ library during the test using ctypes.

**Reply by RonnyPfannschmidt** [reply]
if you manage actual resources in c, use a context manager, anything else is bad practice and at best a hack

the garbage collector is not a resource manager, pretending it is, and making it work with g collect in tests, will in turn cause your library to be a absolute timebomb, its just asking for a horrendous error to happen

**Comment by nicoddemus** [comment]
I suggest to use an autouse fixture to ensure `gc.collect()` is called for each test:

```python
@pytest.fixture(autouse=True)
def ensure_gc():
    gc.collect()
```

**Reply by Vert53** [reply]
The default (and most granular) scope is "function". Doesn't that mean that this fixture will not be called for each instance of the same test function? But rather once and then shared between the instances?

**Reply by nicoddemus** [reply]
It will be called before each test function is called, even parametrized ones.

**Reply by asottile** [reply]
running `gc.collect` like this does not mirror how python actually works though, so you're testing an environment which is unrealistic

**Comment by Vert53** [comment]
Thanks for the advice everyone

**Comment by madhavajay** [comment]
![tenor](https://user-images.githubusercontent.com/2882739/103352385-d9685800-4af1-11eb-89c1-d6c78ec4a738.gif)
